File: backend/shared/damage_attribution.py
from __future__ import annotations

import re
from typing import Any

from modules.analytics.infrastructure.reference_data import (
    UNKNOWN_NAME,
    find_ability,
    resolve_ability_icon,
    resolve_ability_name,
    resolve_gear_name,
    resolve_melee_weapon_id,
    resolve_weapon_name,
    weapons_by_uuid,
)


def _clean_id(value: Any) -> str:
    text = str(value or "").strip()
    return text or "UNKNOWN"


def _readable_unknown(value: Any, fallback: str) -> str:
    text = str(value or "").strip()
    if not text:
        return fallback
    text = re.sub(r"[_-]+", " ", text)
    return " ".join(part.capitalize() for part in text.split()) or fallback


def resolve_damage_source(
    kill: dict[str, Any],
    *,
    fallback_weapon_id: str | None = None,
    killer_agent_id: str | None = None,
) -> dict[str, Any]:
    finishing_damage = kill.get("finishingDamage") or {}
    damage_type = str(finishing_damage.get("damageType") or "").strip()
    damage_type_lower = damage_type.lower()
    damage_item = _clean_id(
        finishing_damage.get("damageItem") or finishing_damage.get("item")
    )
    fallback_id = _clean_id(fallback_weapon_id)
    source_id = damage_item if damage_item != "UNKNOWN" else fallback_id

    if damage_type_lower == "melee":
        melee_id = resolve_melee_weapon_id()
        return {
            "source_id": melee_id,
            "source_name": "Melee",
            "source_type": "melee",
            "icon": None,
            "damage_type": damage_type or None,
            "is_ability": False,
        }

    if damage_type_lower in {"fall", "falling"}:
        return {
            "source_id": "FALL",
            "source_name": "Caida",
            "source_type": "fall",
            "icon": None,
            "damage_type": damage_type or None,
            "is_ability": False,
        }

    if damage_type_lower in {"bomb", "spike"}:
        return {
            "source_id": "BOMB",
            "source_name": "Spike",
            "source_type": "bomb",
            "icon": None,
            "damage_type": damage_type or None,
            "is_ability": False,
        }

    weapon_doc = weapons_by_uuid().get(source_id)
    ability = find_ability(source_id, agent_id=killer_agent_id)
    if weapon_doc:
        weapon_name = resolve_weapon_name(source_id)
        if weapon_name == UNKNOWN_NAME:
            weapon_name = resolve_gear_name(source_id)
        return {
            "source_id": source_id,
            "source_name": weapon_name if weapon_name != UNKNOWN_NAME else _readable_unknown(source_id, "Arma desconocida"),
            "source_type": "weapon",
            "icon": weapon_doc.get("displayIcon") if weapon_doc else None,
            "damage_type": damage_type or None,
            "is_ability": False,
        }

    if damage_type_lower == "ability" or ability:
        ability_id = str((ability or {}).get("uuid") or source_id)
        ability_name = (
            (ability or {}).get("displayName")
            or resolve_ability_name(ability_id)
            or _readable_unknown(source_id, "Habilidad desconocida")
        )
        ability_icon = (ability or {}).get("displayIcon") or resolve_ability_icon(ability_id)
        return {
            "source_id": ability_id,
            "source_name": ability_name if ability_name != UNKNOWN_NAME else _readable_unknown(source_id, "Habilidad desconocida"),
            "source_type": "ability",
            "icon": ability_icon,
            "damage_type": damage_type or None,
            "is_ability": True,
        }

    if damage_type_lower == "weapon":
        return {
            "source_id": source_id,
            "source_name": _readable_unknown(source_id, "Arma desconocida"),
            "source_type": "weapon",
            "icon": None,
            "damage_type": damage_type or None,
            "is_ability": False,
        }

    return {
        "source_id": source_id,
        "source_name": _readable_unknown(source_id, damage_type or "Unknown"),
        "source_type": damage_type_lower if damage_type_lower in {"melee", "fall", "bomb"} else "unknown",
        "icon": None,
        "damage_type": damage_type or None,
        "is_ability": False,
    }

File: frontend/src/utils/damageAttribution.ts
import type { AgentContent } from "../types/agents";
import type { RawKillEvent, RawPlayer } from "../types/matches";

export type DamageSourceType =
  | "weapon"
  | "ability"
  | "melee"
  | "fall"
  | "bomb"
  | "unknown";

export type WeaponCatalogEntry = {
  id: string;
  displayName: string;
  displayIcon?: string | null;
};

export type DamageSourceResolution = {
  id: string;
  name: string;
  icon?: string | null;
  type: DamageSourceType;
  isAbility: boolean;
};

type ResolvedAbility = {
  id: string;
  name: string;
  icon?: string | null;
};

const CONNECTIONLESS_ABILITY_NAMES_BY_AGENT: Record<string, Set<string>> = {
  breach: new Set(["replica"]),
  brimstone: new Set(["incendiaria", "incendiario", "ataqueorbital", "golpeorbital"]),
  cypher: new Set(["cabletrampa"]),
  gekko: new Set(["pogobrutal", "mosh"]),
  killjoy: new Set(["nanoplaga", "nanoenjambre", "torreta"]),
  kayo: new Set(["fragmentacion", "fragmento", "frag"]),
  phoenix: new Set(["manitascalientes", "combustion", "muroabrasador", "llamarada"]),
  raze: new Set(["paqueteexclusivo", "fardoexplosivo", "carcasasdepinturas", "balasdepintura", "bumbot", "botexplosivo"]),
  skye: new Set(["forjacaminos", "precursor"]),
  sova: new Set(["proyectilelectrico", "flechaexplosiva"]),
  tejo: new Set(["entregaespecial", "envioespecial", "misilescrucero", "descargaguiada", "armagedon"]),
  viper: new Set(["mordedura", "venenodeserpiente"]),
  vyse: new Set(["enredaderafilosa", "zarzallacerante"]),
};

const ABILITY_SLOT_ALIASES: Record<string, string> = {
  grenadeability: "grenade",
  grenade: "grenade",
  ability1: "ability1",
  ability2: "ability2",
  ultimate: "ultimate",
};

const ABILITY_WEAPON_SLOT_ALIASES = new Map<string, string>([
  [
    "22697a3d-45bf-8dd7-4fec-84a9e28c69d7:856d9a7e-4b06-dc37-15dc-9d809c37cb90",
    "ability1",
  ],
  [
    "22697a3d-45bf-8dd7-4fec-84a9e28c69d7:39099fb5-4293-def4-1e09-2e9080ce7456",
    "ultimate",
  ],
  [
    "bb2a4828-46eb-8cd1-e765-15848195d751:95336ae4-45d4-1032-cfaf-6bad01910607",
    "ultimate",
  ],
]);

function cleanId(value?: string | null): string {
  return String(value ?? "").trim();
}

function normalizeLookup(value?: string | null): string {
  return String(value ?? "")
    .trim()
    .toLowerCase()
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .replace(/[^a-z0-9]+/g, "");
}

function abilityIconPath(agentId: string, abilityName: string): string | null {
  if (!agentId || !abilityName) return null;
  const sanitized = abilityName
    .trim()
    .replace(/[<>:"/\\|?*\x00-\x1F]/g, "_")
    .replace(/\s+/g, "_")
    .replace(/_+/g, "_")
    .replace(/^[._]+|[._]+$/g, "");
  return sanitized
    ? `/content/agents/${agentId}/abilities/${sanitized}/displayIcon.png`
    : null;
}

function readableFallback(value: string, fallback: string): string {
  if (!value) return fallback;
  const text = value.replace(/[_-]+/g, " ").trim();
  return text
    ? text
        .split(/\s+/)
        .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
        .join(" ")
    : fallback;
}

function findAbility(
  value: string,
  agentsById: Map<string, AgentContent>,
  agentId?: string,
): ResolvedAbility | null {
  let normalized = normalizeLookup(value);
  const agentKey = cleanId(agentId);
  const weaponSlotAlias = ABILITY_WEAPON_SLOT_ALIASES.get(
    `${agentKey}:${cleanId(value).toLowerCase()}`,
  );
  const slotAlias = ABILITY_SLOT_ALIASES[normalized];
  if (weaponSlotAlias) {
    normalized = weaponSlotAlias;
  } else if (slotAlias) {
    normalized = slotAlias;
  }
  const agents = agentId
    ? [agentsById.get(agentId), ...[...agentsById.values()].filter((agent) => cleanId(agent.uuid ?? agent.id) !== agentId)]
    : [...agentsById.values()];

  for (const agent of agents) {
    if (!agent) continue;
    const currentAgentId = cleanId(agent.uuid ?? agent.id);
    for (const ability of agent.abilities ?? []) {
      const abilityId = cleanId(ability.uuid ?? ability.id);
      const slot = cleanId(ability.slot);
      const name = cleanId(ability.displayName);
      const assetPath = cleanId(ability.assetPath);
      const rawName = cleanId(ability.rawName);
      const keys = [
        abilityId,
        slot,
        name,
        rawName,
        assetPath,
        normalizeLookup(slot),
        currentAgentId && slot ? `${currentAgentId}:${slot}` : "",
        currentAgentId && name ? `${currentAgentId}:${normalizeLookup(name)}` : "",
        currentAgentId && slot ? `${currentAgentId}:${normalizeLookup(slot)}` : "",
      ];
      const matchesDirectly = keys.some(
        (key) => key && normalizeLookup(key) === normalized,
      );
      const matchesAgentSlot =
        currentAgentId === agentKey && normalizeLookup(slot) === normalized;
      if (matchesDirectly || matchesAgentSlot) {
        return {
          id:
            abilityId ||
            assetPath ||
            (currentAgentId && slot
              ? `${currentAgentId}:${slot}`
              : `${currentAgentId}:${normalizeLookup(name)}`),
          name: name || readableFallback(value, "Habilidad desconocida"),
          icon: ability.displayIcon ?? abilityIconPath(currentAgentId, name) ?? null,
        };
      }
    }
  }

  return null;
}

export function resolveKillDamageSource(
  kill: RawKillEvent,
  killer: RawPlayer | undefined,
  weaponsById: Map<string, WeaponCatalogEntry>,
  agentsById: Map<string, AgentContent>,
): DamageSourceResolution {
  const damageType = cleanId(kill.finishingDamage?.damageType);
  const typeLower = damageType.toLowerCase();
  const damageItem = cleanId(
    kill.finishingDamage?.damageItem ?? kill.finishingDamage?.item,
  );
  const killerAgentId = cleanId(killer?.characterId);
  const weapon = damageItem ? weaponsById.get(damageItem) : undefined;
  const ability = damageItem
    ? findAbility(damageItem, agentsById, killerAgentId)
    : null;

  if (typeLower === "melee") {
    return { id: damageItem || "MELEE", name: "Melee", icon: null, type: "melee", isAbility: false };
  }
  if (typeLower === "fall" || typeLower === "falling") {
    return { id: "FALL", name: "Caida", icon: null, type: "fall", isAbility: false };
  }
  if (typeLower === "bomb" || typeLower === "spike") {
    return { id: "BOMB", name: "Spike", icon: null, type: "bomb", isAbility: false };
  }
  if (weapon) {
    return {
      id: damageItem || "UNKNOWN",
      name: weapon?.displayName ?? readableFallback(damageItem, "Arma desconocida"),
      icon: weapon?.displayIcon ?? null,
      type: "weapon",
      isAbility: false,
    };
  }

  if (typeLower === "ability" || ability) {
    return {
      id: ability?.id ?? damageItem ?? "ABILITY_UNKNOWN",
      name: ability?.name ?? readableFallback(damageItem, "Habilidad desconocida"),
      icon: ability?.icon ?? null,
      type: "ability",
      isAbility: true,
    };
  }

  if (typeLower === "weapon") {
    return {
      id: damageItem || "UNKNOWN",
      name: readableFallback(damageItem, "Arma desconocida"),
      icon: null,
      type: "weapon",
      isAbility: false,
    };
  }

  return {
    id: damageItem || "UNKNOWN",
    name: readableFallback(damageItem, damageType || "Unknown"),
    icon: null,
    type: "unknown",
    isAbility: false,
  };
}

export function shouldSuppressKillConnectionLine(
  source: DamageSourceResolution,
  killerAgent?: AgentContent | null,
): boolean {
  if (!source.isAbility && source.type !== "ability") return false;
  const agentName = normalizeLookup(killerAgent?.displayName ?? killerAgent?.name);
  const abilityName = normalizeLookup(source.name);
  if (!agentName || !abilityName) return false;
  return CONNECTIONLESS_ABILITY_NAMES_BY_AGENT[agentName]?.has(abilityName) ?? false;
}
